File: answers/lintcode/Binary Tree Preorder Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in vector which contains node values.
     */
    vector<int> preorderTraversal(TreeNode *root) {
        // write your code here
        vector<int> traversal;
        preorderTraversal(root, traversal);
        return traversal;
    }
    void preorderTraversal(TreeNode *root, std::vector<int> &traversal) {
    	if (NULL == root) {
    		return;
    	}
    	traversal.push_back(root->val);
    	preorderTraversal(root->left, traversal);
    	preorderTraversal(root->right, traversal);
    }
};

File: answers/leetcode/Binary Tree Preorder Traversal/Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> traversal = new ArrayList<Integer>();
        this.dfs(root, traversal);
        return traversal;
    }
    
    private void dfs(TreeNode root, List<Integer> traversal) {
        if (null == root) {
            return;
        }
        traversal.add(root.val);
        this.dfs(root.left, traversal);
        this.dfs(root.right, traversal);
    }
}